From 690e0fb418d26a949970ef419ce5d04b911d51f8 Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Fri, 24 Nov 2017 10:29:12 +0200
Subject: [PATCH 179/741] sdk_dpaa: ceetm: add independent CEETM congestion
 thresholds

commit c46e8f98b01603be168a96eba80ee46164fa05e4 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Configure the CEETM egress congestion thresholds independently from the
default Ethernet driver's Work Queues. Allow the user to edit the
thresholds through menuconfig.

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 .../net/ethernet/freescale/sdk_dpaa/Kconfig   | 23 +++++++++++++++++++
 .../freescale/sdk_dpaa/dpaa_eth_ceetm.c       |  4 ++--
 2 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/freescale/sdk_dpaa/Kconfig b/drivers/net/ethernet/freescale/sdk_dpaa/Kconfig
index 92118b76b08e..b21d23623e59 100644
--- a/drivers/net/ethernet/freescale/sdk_dpaa/Kconfig
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/Kconfig
@@ -20,6 +20,29 @@ config FSL_DPAA_CEETM
 	help
 	Enable QoS offloading support through the CEETM hardware block.
 
+config FSL_DPAA_CEETM_CCS_THRESHOLD_1G
+	hex "CEETM egress congestion threshold on 1G ports"
+	depends on FSL_DPAA_CEETM
+	range 0x1000 0x10000000
+	default "0x000a0000"
+	help
+	  The size in bytes of the CEETM egress Class Congestion State threshold on 1G ports.
+	  The threshold needs to be configured keeping in mind the following factors:
+	       - A threshold too large will buffer frames for a long time in the TX queues,
+	       when a small shaping rate is configured. This will cause buffer pool depletion
+	       or out of memory errors. This in turn will cause frame loss on RX;
+	       - A threshold too small will cause unnecessary frame loss by entering
+	       congestion too often.
+
+config FSL_DPAA_CEETM_CCS_THRESHOLD_10G
+	hex "CEETM egress congestion threshold on 10G ports"
+	depends on FSL_DPAA_CEETM
+	range 0x1000 0x20000000
+	default "0x00640000"
+	help
+	  The size in bytes of the CEETM egress Class Congestion State threshold on 10G ports.
+	  See FSL_DPAA_CEETM_CCS_THRESHOLD_1G for details.
+
 config FSL_DPAA_OFFLINE_PORTS
 	bool "Offline Ports support"
 	depends on FSL_SDK_DPAA_ETH
diff --git a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c
index cac613b769a5..bf635b9d0277 100644
--- a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c
@@ -185,9 +185,9 @@ static int ceetm_config_ccg(struct qm_ceetm_ccg **ccg,
 
 	/* Set the congestion state thresholds according to the link speed */
 	if (dpa_priv->mac_dev->if_support & SUPPORTED_10000baseT_Full)
-		cs_th = CONFIG_FSL_DPAA_CS_THRESHOLD_10G;
+		cs_th = CONFIG_FSL_DPAA_CEETM_CCS_THRESHOLD_10G;
 	else
-		cs_th = CONFIG_FSL_DPAA_CS_THRESHOLD_1G;
+		cs_th = CONFIG_FSL_DPAA_CEETM_CCS_THRESHOLD_1G;
 
 	qm_cgr_cs_thres_set64(&ccg_params.cs_thres_in, cs_th, 1);
 	qm_cgr_cs_thres_set64(&ccg_params.cs_thres_out,
-- 
2.17.1

